## A jupyter notebook version file for the `main.py`


Set `autoreload` to execute the change in `.py` files.


In [1]:
%load_ext autoreload
%autoreload 2

In [2]:
import _pickle
import natsort
from pathlib import Path
import os

os.environ["CUDA_VISIBLE_DEVICES"] = "0"

dataset_dir = "D:/Databases/SAMM_longvideos"
# dataset_dir = "F:/HEH/Databases/SAMM_longvideos"
# dataset_dir = "/data/disk1/heh/databases/SAMM_longvideos"

images_loading = False
image_size = 128
load_cropped_images = False
# expression_type = "mae"  # macro-expression spotting
expression_type = "me"  # micro-expression spotting
save_x = False
debug_preds = True
labeling_function = "pseudo_labeling"
# labeling_function = "original_labeling"
model_names = {
    0: "SOFTNet",
    1: "SOFTNetCBAM",
    2: "ViT-B",
    3: "SL-ViT-B",
    4: "Swin-T",
    5: "Swin-S",
    6: "L-Swin-T",
    7: "S-Swin-T",
    8: "SL-Swin-T",
    9: "SL-Swin-S",
}
model_name = model_names[2]
batch_size = 48
epochs = 25
save_preds = False
preds_stem = (
    f"{expression_type}_"
    + model_name.lower().replace("-", "_")
    + f"_batch_size_{batch_size}"
    + f"_epochs_{epochs}"
    + f"_{labeling_function}"
    + f"_{image_size}"
    # + "_2"
)
preds_path = Path(dataset_dir, "preds", preds_stem).with_suffix(".pkl")
print(f"preds_path: {preds_path}")


preds_path: D:\Databases\SAMM_longvideos\preds\me_vit_b_batch_size_48_epochs_25_pseudo_labeling_128.pkl


## Crop Images


In [3]:
from image_processing import *

# crop_images_dev(dataset_dir)


## Load Images


When debug the image processing, the videos_images is from cropped_rawpic, whereas the other variables are from rawpic.


In [4]:
from image_processing import *

# videos_images, subjects, subjects_videos_code = load_images(dataset_dir)

videos_images, subjects, subjects_videos_code = load_images_dev(
    dataset_dir,
    images_loading=images_loading,
    image_size=image_size,
    load_cropped_images=load_cropped_images,
)


subject_video: 006_1
subject_video: 006_2
subject_video: 006_3
subject_video: 006_4
subject_video: 006_5
subject_video: 006_6
subject_video: 006_7
subject_video: 007_3
subject_video: 007_4
subject_video: 007_5
subject_video: 007_6
subject_video: 007_7
subject_video: 008_1
subject_video: 008_5
subject_video: 008_6
subject_video: 008_7
subject_video: 009_2
subject_video: 009_3
subject_video: 009_4
subject_video: 009_6
subject_video: 009_7
subject_video: 010_1
subject_video: 010_2
subject_video: 010_3
subject_video: 010_4
subject_video: 010_5
subject_video: 010_6
subject_video: 010_7
subject_video: 011_1
subject_video: 011_2
subject_video: 011_3
subject_video: 011_4
subject_video: 011_5
subject_video: 011_6
subject_video: 011_7
subject_video: 012_3
subject_video: 012_4
subject_video: 012_5
subject_video: 012_6
subject_video: 012_7
subject_video: 013_1
subject_video: 013_2
subject_video: 013_3
subject_video: 013_6
subject_video: 013_7
subject_video: 014_1
subject_video: 014_2
subject_video

In [5]:
print("subjects:", subjects)
print("subjects_videos_code:", subjects_videos_code)


subjects: ['006', '007', '008', '009', '010', '011', '012', '013', '014', '015', '016', '017', '018', '019', '020', '021', '022', '023', '024', '025', '026', '028', '030', '031', '032', '033', '034', '035', '036', '037']
subjects_videos_code: [['1', '2', '3', '4', '5', '6', '7'], ['3', '4', '5', '6', '7'], ['1', '5', '6', '7'], ['2', '3', '4', '6', '7'], ['1', '2', '3', '4', '5', '6', '7'], ['1', '2', '3', '4', '5', '6', '7'], ['3', '4', '5', '6', '7'], ['1', '2', '3', '6', '7'], ['1', '2', '3', '4', '5', '6', '7'], ['1', '3', '5', '6', '7'], ['1', '2', '4', '5', '6', '7'], ['1', '2', '3', '4', '5', '6'], ['1', '2', '3', '4', '5', '6', '7'], ['1', '2', '3', '4', '5', '7'], ['1', '2', '3', '4', '5', '6', '7'], ['3', '7'], ['2', '3', '4', '5', '6'], ['1', '4'], ['2', '3', '5'], ['3', '4', '5', '6'], ['1', '2', '3', '5', '6', '7'], ['4'], ['1', '2', '5'], ['3'], ['2', '3', '4', '5', '6'], ['1', '2', '3', '4', '5', '6', '7'], ['3', '6', '7'], ['1', '2', '3', '4', '5', '6', '7'], ['2', '4',

## Load Excel 


In [6]:
from label_processing import load_excel

Excel_data = load_excel(dataset_dir)


Data Columns: Index(['subject', 'Filename', 'video_code', 'onset', 'apex', 'offset',
       'Duration', 'expression_type', 'Action Units', 'Notes',
       'subject_video_code', 'subject_code'],
      dtype='object')


In [7]:
Excel_data.head(5)


subject Filename video_code onset  apex offset Duration expression_type   
0     006  006_1_1          1   566   648    743      178           Macro  \
1     006  006_1_2          1  3562  3588   3632       71     Micro - 1/2   
2     006  006_1_3          1  1912  1948   1988       77     Micro - 1/2   
3     006  006_1_4          1   324   368    403       80     Micro - 1/2   
4     006  006_1_5          1  3343  3388   3424       82     Micro - 1/2   

  Action Units           Notes subject_video_code subject_code  
0    4(B/C)+7B             NaN              006_1          006  
1          4+7             NaN              006_1          006  
2            4  While blinking              006_1          006  
3          4+7             NaN              006_1          006  
4          4+7             NaN              006_1          006

## Load Ground Truth Labels


In [8]:
from label_processing import load_ground_truth_labels


(
    clean_videos_images,
    clean_subjects_videos_code,
    clean_subjects,
    clean_subjects_videos_ground_truth_labels,
) = load_ground_truth_labels(
    dataset_dir,
    expression_type,
    videos_images,
    subjects_videos_code,
    subjects,
    Excel_data,
)


required_videos_index:  [0, 1, 2, 4, 7, 8, 9, 10, 11, 16, 17, 22, 24, 28, 29, 30, 31, 32, 33, 34, 35, 39, 40, 44, 45, 46, 47, 49, 50, 51, 54, 62, 65, 68, 69, 71, 73, 75, 78, 79, 80, 82, 85, 88, 90, 91, 92, 93, 94, 96, 98, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 114, 115, 117, 118, 120, 121, 122, 128, 130, 131, 134, 135, 136, 137, 141, 143, 144]
len(clean_videos_images) = 79


In [9]:
print("len(clean_subjects): ", len(clean_subjects))
print("clean_subjects: ", clean_subjects)
print("len(clean_subjects_videos_code): ", len(clean_subjects_videos_code))
print("clean_subjects_videos_codes: ", clean_subjects_videos_code)
print(
    "len(clean_subjects_videos_ground_truth_labels): ",
    len(clean_subjects_videos_ground_truth_labels),
)
print(
    "clean_subjects_videos_ground_truth_labels: ",
    clean_subjects_videos_ground_truth_labels,
)


len(clean_subjects):  29
clean_subjects:  ['006' '007' '009' '010' '011' '012' '013' '014' '015' '016' '017' '018'
 '019' '020' '021' '022' '023' '024' '025' '026' '028' '030' '031' '032'
 '033' '034' '035' '036' '037']
len(clean_subjects_videos_code):  29
clean_subjects_videos_codes:  [['1', '2', '3', '5'], ['3', '4', '5', '6', '7'], ['2', '3'], ['2', '4'], ['1', '2', '3', '4', '5', '6', '7'], ['3', '7'], ['1', '7'], ['1', '2', '3', '5', '6', '7'], ['5'], ['7'], ['3', '6'], ['1', '3', '5', '7'], ['3', '4', '5'], ['1', '4', '7'], ['7'], ['2', '3', '4', '5'], ['1'], ['2'], ['4', '5', '6'], ['1', '2', '3', '5', '6', '7'], ['4'], ['1', '5'], ['3'], ['3', '4', '6'], ['1', '2'], ['3', '7'], ['1', '4', '5', '6', '7'], ['7'], ['3', '4']]
len(clean_subjects_videos_ground_truth_labels):  29
clean_subjects_videos_ground_truth_labels:  [[[[3561, 3631], [1911, 1987], [323, 402], [3342, 3423], [5159, 5258]], [[79, 173]], [[138, 201], [1286, 1355]], [[1135, 1213], [1829, 1925], [4957, 5050]]], [[[24

In [10]:
clean_subjects_videos_code_len = 0
for i, clean_subject_videos_code in enumerate(clean_subjects_videos_code):
    clean_subjects_videos_code_len += len(clean_subject_videos_code)
    print(
        i,
        clean_subjects[i],
        clean_subjects_videos_code[i],
        clean_subjects_videos_code_len,
    )


0 006 ['1', '2', '3', '5'] 4
1 007 ['3', '4', '5', '6', '7'] 9
2 009 ['2', '3'] 11
3 010 ['2', '4'] 13
4 011 ['1', '2', '3', '4', '5', '6', '7'] 20
5 012 ['3', '7'] 22
6 013 ['1', '7'] 24
7 014 ['1', '2', '3', '5', '6', '7'] 30
8 015 ['5'] 31
9 016 ['7'] 32
10 017 ['3', '6'] 34
11 018 ['1', '3', '5', '7'] 38
12 019 ['3', '4', '5'] 41
13 020 ['1', '4', '7'] 44
14 021 ['7'] 45
15 022 ['2', '3', '4', '5'] 49
16 023 ['1'] 50
17 024 ['2'] 51
18 025 ['4', '5', '6'] 54
19 026 ['1', '2', '3', '5', '6', '7'] 60
20 028 ['4'] 61
21 030 ['1', '5'] 63
22 031 ['3'] 64
23 032 ['3', '4', '6'] 67
24 033 ['1', '2'] 69
25 034 ['3', '7'] 71
26 035 ['1', '4', '5', '6', '7'] 76
27 036 ['7'] 77
28 037 ['3', '4'] 79


## Calculate k


In [11]:
from label_processing import calculate_k

k = calculate_k(clean_subjects_videos_ground_truth_labels)


k (Half of average length of expression) =  37


## Extract Features


It takes about 1034 m.


Compute Optical Flow Features (shape = [128, 128, 3])


In [12]:
from features_extraction import extract_features

# clean_videos_images_features = extract_features(clean_videos_images, k, image_size=128)


## Preprocessing


It takes 908 m.


In [13]:
from pre_processing import pre_process

# resampled_clean_videos_images_features = pre_process(
#     clean_videos_images, clean_videos_images_features, k
# )


### Dump Resampled Clean Videos Images Features


In [14]:
# with open(
#     Path(
#         dataset_dir,
#         f"resampled_clean_videos_images_{expression_type}_features_{image_size}",
#     ).with_suffix(".pkl"),
#     "wb",
# ) as pkl_file:
#     _pickle.dump(resampled_clean_videos_images_features, pkl_file)
#     pkl_file.close()


### Load **Original** Resampled Clean Videos Images Features


In [15]:
# with open(
#     Path(
#         dataset_dir,
#         f"original_resampled_clean_videos_images_{expression_type}_features.pkl",
#     ),
#     "rb",
# ) as pkl_file:
#     resampled_clean_videos_images_features = _pickle.load(pkl_file)
#     pkl_file.close()


### Load Resampled Clean Videos Images Features


In [16]:
if save_x is not False and debug_preds is not True:
    with open(
        Path(
            dataset_dir,
            f"resampled_clean_videos_images_{expression_type}_features_{image_size}",
        ).with_suffix(".pkl"),
        "rb",
    ) as pkl_file:
        resampled_clean_videos_images_features = _pickle.load(pkl_file)
        pkl_file.close()


In [17]:
if save_x is not False and debug_preds is not True:
    print(
        "len(resampled_clean_videos_images_features): ",
        len(resampled_clean_videos_images_features),
    )
    print(
        "len(resampled_clean_videos_images_features[0]): ",
        len(resampled_clean_videos_images_features[0]),
    )
    print(
        "resampled_clean_videos_images_features[0][0].shape: ",
        resampled_clean_videos_images_features[0][0].shape,
    )


## Pseudo Labeling


In [18]:
from labeling import *

if labeling_function == "pseudo_labeling":
    labels = get_pseudo_labels(
        clean_videos_images, clean_subjects_videos_ground_truth_labels, k
    )
elif labeling_function == "original_labeling":
    labels = get_original_labels(
        clean_videos_images, clean_subjects_videos_ground_truth_labels, k
    )


Total frames: 310682


## Prepare for LOSO


In [19]:
from loso_preparing import prepare_for_loso_dev

if save_x is False:
    resampled_clean_videos_images_features = None

y, groups = prepare_for_loso_dev(
    resampled_clean_videos_images_features,
    labels,
    clean_subjects,
    clean_videos_images,
    clean_subjects_videos_ground_truth_labels,
    k,
    save_x,
    expression_type,
    dataset_dir,
)


Frame Index for each subject:-

subject 006 ( group = 0): 0 -> 19551
subject 006 has 4 clean video(s)
sum clean_subject_videos_ground_truth_labels_len:  4

subject 007 ( group = 1): 19551 -> 34566
subject 007 has 5 clean video(s)
sum clean_subject_videos_ground_truth_labels_len:  9

subject 009 ( group = 2): 34566 -> 38612
subject 009 has 2 clean video(s)
sum clean_subject_videos_ground_truth_labels_len:  11

subject 010 ( group = 3): 38612 -> 47425
subject 010 has 2 clean video(s)
sum clean_subject_videos_ground_truth_labels_len:  13

subject 011 ( group = 4): 47425 -> 90918
subject 011 has 7 clean video(s)
sum clean_subject_videos_ground_truth_labels_len:  20

subject 012 ( group = 5): 90918 -> 97820
subject 012 has 2 clean video(s)
sum clean_subject_videos_ground_truth_labels_len:  22

subject 013 ( group = 6): 97820 -> 114534
subject 013 has 2 clean video(s)
sum clean_subject_videos_ground_truth_labels_len:  24

subject 014 ( group = 7): 114534 -> 139833
subject 014 has 6 clean vid

## Training


In [20]:
if debug_preds is False:
    from training_dev import train

    preds = train(
        dataset_dir,
        clean_subjects,
        image_size=image_size,
        y=y,
        expression_type=expression_type,
        model_name=model_name,
        train_or_not=True,
        batch_size=batch_size,
        epochs=epochs,
        preds_path=preds_path,
    )
else:
    with open(preds_path, "rb") as pkl_file:
        preds = _pickle.load(pkl_file)
        pkl_file.close()


In [21]:
if save_preds is True:
    with open(preds_path, "wb") as pkl_file:
        _pickle.dump(preds, pkl_file)
        pkl_file.close()


## Spotting and Evaluation


In [22]:
from functions import *

metric_fn, result_dict = spot_and_evaluate(
    preds,
    clean_subjects_videos_ground_truth_labels,
    clean_videos_images,
    clean_subjects,
    clean_subjects_videos_code,
    k,
    p=0.60,
    show_plot_or_not=False,
)


Split 1/29 is in process.
0 video(s) have been processed.
The current video be processed: subject 006, video 1
The current video be processed: subject 006, video 2
The current video be processed: subject 006, video 3
The current video be processed: subject 006, video 5

True Positive: 1, False Posive: 42, False Negative: 10
Precision = 0.023255813953488594, Recall =0.09090909090909113, F1-Score = 0.037037037037037555
Split 1/29 is processed.

Split 2/29 is in process.
4 video(s) have been processed.
The current video be processed: subject 007, video 3
The current video be processed: subject 007, video 4
The current video be processed: subject 007, video 5
The current video be processed: subject 007, video 6
The current video be processed: subject 007, video 7

True Positive: 3, False Posive: 63, False Negative: 18
Precision = 0.04545454545454568, Recall =0.14285714285714307, F1-Score = 0.06896551724137981
Split 2/29 is processed.

Split 3/29 is in process.
9 video(s) have been processe

## Final Evaluation


In [23]:
from evaluation import final_evaluate

final_evaluate(metric_fn, result_dict)


True Positive: 35, False Posive: 496, False Negative: 124
Final Precision = 0.06591337099811698,
Final Recall =0.22012578616352224,
Final F1-Score = 0.10144927536231936

Highest Precision = 0.07487922705314032,
Highest Recall =0.24218750000000022,
Highest F1-Score = 0.11439114391143962


| Parameters | Value | Value 
| --- | --- | ---
| model | 3D-CNN | SOFTNet 
| epochs | | 
| batch_size | | 48  
| learning_rate | | 0.0005
| True Positive | | 38
| False Positive | | 303
| False Negative | | 121
| Precision | | 0.1114
| Recall | | 0.3188 | 0.2309
| F1-Score | 0.0466 | 0.1520


| Parameters | Value | Value | Value | Value
| --- | --- | --- | --- | ---
| model | Swin-T | Swin-T | SL-Swin-T | SL-Swin-T_2
| epochs | 25 (4350 m) | 25 (4350 m) | 25 (7399 m) | 25 (7400 m)
| batch_size | 48 | 48 | 48 | 48
| p | 0.57 | 0.60 | 0.59 | 0.60
| True Positive | 37 | 29 | 26 | 33
| False Positive | 518 | 468 | 409 | 440
| False Negative | 122 | 130 | 133 | 126
| Precision | 0.0667 | 0.0583 | 0.0597 | 0.0698
| Recall | 0.2327 | 0.1823 | 0.1635 | 0.2075
| F1-Score | 0.1036 | 0.0884 | 0.0875 | 0.1044


| Parameters | Value | Value | Value | Value
| --- | --- | --- | --- | ---
| model | ViT-B | SL-ViT | S-Swin-T | L-Swin-T
| epochs | 25 (15605 m) | 25 ( m) | 25 (4827 m) | 25 (6960 m)
| batch_size | 48 | 48 | 48 | 48
| p | 0.57 | 0.60 | 0.59 | 0.60
| True Positive |  |  |  | 
| False Positive |  |  |  | 
| False Negative |  |  |  | 
| Precision |  |  |  | 
| Recall |  |  |  | 
| F1-Score |  |  |  | 


## Ablation Study


In [24]:
from functions import *

ablation_dict = ablation_study_p_dev(
    preds,
    clean_subjects_videos_ground_truth_labels,
    clean_videos_images,
    clean_subjects,
    clean_subjects_videos_code,
    k,
)


 p | TP | FP | FN | Precision | Recall | F1-Score
0.01 | 94 | 2571 | 65 | 0.0353 | 0.5912 | 0.0666 |
0.02 | 94 | 2487 | 65 | 0.0364 | 0.5912 | 0.0686 |
0.03 | 91 | 2403 | 68 | 0.0365 | 0.5723 | 0.0686 |
0.04 | 89 | 2316 | 70 | 0.0370 | 0.5597 | 0.0694 |
0.05 | 89 | 2253 | 70 | 0.0380 | 0.5597 | 0.0712 |
0.06 | 89 | 2188 | 70 | 0.0391 | 0.5597 | 0.0731 |
0.07 | 89 | 2121 | 70 | 0.0403 | 0.5597 | 0.0751 |
0.08 | 88 | 2057 | 71 | 0.0410 | 0.5535 | 0.0764 |
0.09 | 87 | 1995 | 72 | 0.0418 | 0.5472 | 0.0776 |
0.10 | 84 | 1924 | 75 | 0.0418 | 0.5283 | 0.0775 |
0.11 | 84 | 1862 | 75 | 0.0432 | 0.5283 | 0.0798 |
0.12 | 82 | 1806 | 77 | 0.0434 | 0.5157 | 0.0801 |
0.13 | 82 | 1756 | 77 | 0.0446 | 0.5157 | 0.0821 |
0.14 | 80 | 1697 | 79 | 0.0450 | 0.5031 | 0.0826 |
0.15 | 78 | 1656 | 81 | 0.0450 | 0.4906 | 0.0824 |
0.16 | 76 | 1615 | 83 | 0.0449 | 0.4780 | 0.0822 |
0.17 | 73 | 1572 | 86 | 0.0444 | 0.4591 | 0.0809 |
0.18 | 72 | 1537 | 87 | 0.0447 | 0.4528 | 0.0814 |
0.19 | 70 | 1508 | 89 | 0.0444 |